src/optimizer/query_graph.c

```diff
@@ -1518,6 +1522,42 @@ qo_insert_segment (QO_NODE * head, QO_NODE * tail, PT_NODE * node, QO_ENV * env,
 
   bitset_add (&(QO_NODE_SEGS (head)), QO_SEG_IDX (seg));
 
+  /* check is_not_null */
+  entity = QO_NODE_ENTITY_SPEC (head);
+  if (pt_is_name_node (node) && PT_SPEC_IS_ENTITY (entity))
+    {
+      int i;
+      bool found = false;
+      cls = sm_find_class (entity->info.spec.entity_name->info.name.original);
+      constraints = sm_class_constraints (cls);
+
+      while (constraints != NULL)
+	{
+	  if (!SM_IS_CONSTRAINT_NOT_NULL_FAMILY (constraints->type))
+	    {
+	      constraints = constraints->next;
+	      continue;
+	    }
+
+	  /* check columns on this constraint */
+	  for (i = 0; constraints->attributes[i]; i++)
+	    {
+	      attrp = constraints->attributes[i];
+	      if (intl_identifier_casecmp (node->info.name.original, attrp->header.name) == 0)
+		{
+		  QO_SEG_IS_NOT_NULL (seg) = true;
+		  found = true;
+		  break;
+		}
+	    }
+	  if (found)
+	    {
+	      break;
+	    }
+	  constraints = constraints->next;
+	}
+    }
+
   env->nsegs++;
```

Comment:
Same content, but I think the code could be shorter

```suggestion
  env->nsegs++;
    
    /* check is_not_null */
  entity = QO_NODE_ENTITY_SPEC (head);
  if (pt_is_name_node (node) && PT_SPEC_IS_ENTITY (entity))
    {
      int i;
      cls = sm_find_class (entity->info.spec.entity_name->info.name.original);
      constraints = sm_class_constraints (cls);

      while (constraints != NULL)
	{
	  if (SM_IS_CONSTRAINT_NOT_NULL_FAMILY (constraints->type))
	    {	      
	       /* check columns on this constraint */
	      for (i = 0; constraints->attributes[i]; i++)
	        {
	          attrp = constraints->attributes[i];
	          if (intl_identifier_casecmp (node->info.name.original, attrp->header.name) == 0)
		    {
		      QO_SEG_IS_NOT_NULL (seg) = true;
		     return seg;
		    }
	        }
	    }
	  constraints = constraints->next;
	}
    }
```

Reply:
Thanks for your comments. But I don't want to put 'return' in the middle. I'll keep my style.